### ISSUE TITLE
aws_s3_bucket lifecycle_rule should support filter

### ISSUE THREAD

**Issue by kimor79** [hashicorp/terraform-provider-aws#21953]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

AWS recently announced support for filtering objects by size in an S3 bucket lifecycle policy. It would be nice if this was also available in the `aws_s3_bucket` resource.

Example from AWS documentation:
```
<LifecycleConfiguration>
    <Rule>
        <Filter>
            <ObjectSizeGreaterThan>500</ObjectSizeGreaterThan>   
        </Filter>
        <Status>Enabled</Status>
        transition/expiration actions.
    </Rule>
</LifecycleConfiguration>
```

### New or Affected Resource(s)

* aws_s3_bucket

### Potential Terraform Configuration

```hcl
# Copy-paste your Terraform configurations here - for large Terraform configs,
# please use a service like Dropbox and share a link to the ZIP file. For
# security, you can also encrypt the files using our GPG public key.

resource "aws_s3_bucket" "my-bucket" {
  ...

  lifecycle_rule {
    id = "archive"
    enabled = true

    transition {
      days = 90
      storage_class = "GLACIER"
    }

    filter {
      object_size_greater_than = 128000
    }
  }
}
```

### References

* https://aws.amazon.com/about-aws/whats-new/2021/11/amazon-s3-lifecycle-storage-cost-savings/
* https://docs.aws.amazon.com/AmazonS3/latest/userguide/intro-lifecycle-rules.html#intro-lifecycle-rules-filter

**Comment by visit1985** [comment]
Maybe we can add the new `versions` action parameters for `noncurrent_version_transition` and `noncurrent_version_expiration` from the same announcement in the same effort.

### Potential Terraform Configuration

```
resource "aws_s3_bucket" "my-bucket" {
  ...

  lifecycle_rule {
    id = "archive"
    enabled = true

    noncurrent_version_transition {
      storage_class = "STANDARD_IA"
      versions = 4
    }

    noncurrent_version_expiration {
      versions = 4
    }
  }
}
```

### References

- https://docs.aws.amazon.com/AmazonS3/latest/API/API_NoncurrentVersionTransition.html#AmazonS3-Type-NoncurrentVersionTransition-NewerNoncurrentVersions
- https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutBucketLifecycleConfiguration.html

**Comment by dhermes** [comment]
I was just about to file an issue requesting `NewerNoncurrentVersions` support as well, thanks @visit1985

**Comment by pda** [comment]
I see `object_size_greater_than` and `object_size_less_than` filters are supported in the `aws_s3_bucket_lifecycle_configuration` resource type provided by v4.x of hashicorp/terraform-provider-aws.

Landed in d8640c1b79bfa4f369a40b5a8d49a53f2fdcf3c9 / #22579 / [v4.0.0](https://github.com/hashicorp/terraform-provider-aws/releases/tag/v4.0.0) (11 Feb 2022):

> resource/aws_s3_bucket: The `lifecycle_rule` argument has been deprecated and is now read-only. Use the `aws_s3_bucket_lifecycle_configuration` resource instead. (https://github.com/hashicorp/terraform-provider-aws/issues/22581)

Documentation:
https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/s3_bucket_lifecycle_configuration#specifying-a-filter-based-on-object-size

I imagine there's no intention to backport it onto the v3.x `aws_s3_bucket` resource, so maybe this issue should close?

A bit more context:

- https://github.com/hashicorp/terraform-provider-aws/issues/23106
- https://www.hashicorp.com/blog/terraform-aws-provider-4-0-refactors-s3-bucket-resource
- https://registry.terraform.io/providers/hashicorp/aws/latest/docs/guides/version-4-upgrade
